engine/lib/phx/src/input/devices/gamepad/state.rs

```diff
@@ -0,0 +1,254 @@
+            name: name.into(),
+            control_state: Default::default(),
+            button_state: Default::default(),
+            axis_state: Default::default(),
+        }
+    }
+
+    pub fn control_state(&self) -> &ControlState {
+        &self.control_state
+    }
+
+    pub fn control_state_mut(&mut self) -> &mut ControlState {
+        &mut self.control_state
+    }
+
+    pub fn reset(&mut self) {
+        self.button_state.reset();
+        self.axis_state.reset();
+    }
+}
+
+pub struct GamepadState {
+    gilrs: Gilrs,
+    device_state: IndexMap<GamepadId, GamepadDeviceState>,
+}
+
+impl Default for GamepadState {
+    fn default() -> Self {
+        let gilrs = GilrsBuilder::new()
+            .add_included_mappings(true) // Load the latest mappings from GitHub
+            .build()
+            .expect("Cannot create Gilrs");
+
+        Self {
+            gilrs,
+            device_state: Default::default(),
+        }
+    }
+}
+
+impl GamepadState {
+    pub fn reset(&mut self) {
+        self.device_state
+            .iter_mut()
+            .for_each(|(_, state)| state.reset());
+    }
+
+    pub fn update(&mut self) -> Option<GamepadId> {
+        let mut res = None;
+
+        while let Some(gilrs_event) = self
+            .gilrs
+            .next_event()
+            .filter_ev(&axis_dpad_to_button, &mut self.gilrs)
+        {
+            self.gilrs.update(&gilrs_event);
+
+            let gamepad_id = gilrs_event.id;
+
+            match gilrs_event.event {
+                EventType::Connected => {
+                    if let Some(state) = self.device_state.get_mut(&gamepad_id) {
+                        state.reset();
+                        state.control_state_mut().connect();
+                    } else {
+                        let pad = self.gilrs.gamepad(gilrs_event.id);
+                        let device_state = GamepadDeviceState::new(pad.name());
+
+                        self.device_state.insert(gamepad_id, device_state);
+                    }
+                }
+                EventType::Disconnected => {
+                    if let Some(state) = self.device_state.get_mut(&gamepad_id) {
+                        state.control_state_mut().disconnect();
+                    } else {
+                        // TODO: warning?
+                    }
+                }
+                EventType::ButtonChanged(gilrs_button, raw_value, _) => {
+                    if let Some(button) = convert_button(gilrs_button) {
+                        if let Some(state) = self.device_state.get_mut(&gamepad_id) {
+                            state.button_state.update(button as _, raw_value != 0.0);
+                            state.axis_state.update(button as _, raw_value);
+
+                            if state.control_state.update() {
+                                res = Some(gamepad_id);
+                            }
+                        }
+
+                        // TODO: threshold check?
```

Comment:
Do we still need this commented code? If so, lets make sure there's an issue on GitHub about it.

Reply:
Leftover from experiments. Removed